extensions/ql-vscode/src/codeql-task-provider.ts

```diff
@@ -0,0 +1,97 @@
+import * as vscode from 'vscode';
+import { CodeQLCliServer } from './cli';
+
+interface CodeQLTaskDefinition extends vscode.TaskDefinition {
+  /**
+   * The command to run.
+   */
+  command: string;
+
+  /**
+   * Additional command arguments.
+   */
+  commandArgs?: string[];
+}
+
+export class CodeQLTaskProvider implements vscode.TaskProvider {
+  // Matches the label in package.json
+  static CodeQLType = 'codeql';
+  private tasks: vscode.Task[] | undefined;
+  private sharedState: string | undefined;
+
+  constructor(private workspaceRoot: string, private cliServer: CodeQLCliServer) { }
+
+  public async provideTasks(): Promise<vscode.Task[]> {
+    return this.getTasks();
+  }
+
+  public resolveTask(_task: vscode.Task): vscode.Task | undefined {
+    const command: string = _task.definition.command;
+    if (command) {
+      const definition: CodeQLTaskDefinition = <any>_task.definition;
+      return this.getTask(definition.command, definition.commandArgs ? definition.commandArgs : [], definition);
+    }
+    return undefined;
+  }
+
+  private getTasks(): vscode.Task[] {
+    if (this.tasks !== undefined) {
+      return this.tasks;
+    }
+    const commands: string[] = ['resolve languages', 'version'];
+    // Each command can have some commandArgs.
+    const commandArgs: string[][] = [[]];
+
+    this.tasks = [];
+    commands.forEach(command => {
+      commandArgs.forEach(commandArgsGroup => {
+        this.tasks!.push(this.getTask(command, commandArgsGroup));
+      });
+    });
```

Comment:
nit: Trick to avoid the type assertion:

```suggestion
    const tasks = [];
    commands.forEach(command => {
      commandArgs.forEach(commandArgsGroup => {
        tasks.push(this.getTask(command, commandArgsGroup));
      });
    });
    this.tasks = tasks;
```